# [ 6. 사용자 정의 오류보다는 표준 오류를 사용하라 ]
https://marchbreeze.notion.site/6-85637b79d0b44411bb7b1d3eccc7a02e?pvs=4

- 예외 처리 방법으로 require, check, assert 함수를 사용하면, 대부분의 코틀린 오류를 처리할 수 있지만, 이외에도 예측하지 못한 상황을 나타내야 하는 경우가 존재함

- 가능하다면, 직접 오류를 정의하는 것보다는 최대한 표준 라이브러리의 오류를 사용하는 것이 좋음

  → 표준 라이브러리의 오류는 많은 개발자가 알고 있으므로 이를 재사용하기 용이함


- 일반적으로 사용되는 예외 :
    1. `IllegalArgumentException`
        - 함수나 메서드에 전달된 인자가 유효하지 않을 때 발생하는 예외
        - 함수의 계약(Contract)을 위반하는 인자가 전달될 때 이 예외가 발생함
        - ex. require 함수를 사용해 특정 조건을 만족하지 않는 경우 발생

            ```kotlin
            fun setAge(age: Int) {
                require(age >= 0) { "Age cannot be negative." }
                // age 설정 로직
            }
            ```


    1. `IllegalStateException`
        - 객체의 상태가 함수나 메서드를 호출하기에 적절하지 않을 때 발생하는 예외
        - 주로 객체의 메서드가 특정 상태에서만 호출 가능할 때 사용됨
        - ex. check 함수를 사용해 객체 상태가 유효하지 않는 경우 발생
            
            ```kotlin
            fun performAction() {
                check(isReady) { "Cannot perform action when the object is not ready." }
                // 액션 수행 로직
            }
            ```
            
    
    1. `IndexOutOfBoundsException`
        - 인덱스 파라미터의 값이 범위를 벗어났다는 예외
        - 컬렉션이나 배열에서 유효하지 않은 인덱스에 접근하려고 할 때 발생
            
            ```kotlin
            val list = listOf(1, 2, 3)
            val element = list[5]  // IndexOutOfBoundsException 발생
            ```


1. `ConcureentModificationException`
    - 한 스레드에서 컬렉션을 순회하는 도중, 해당 컬렉션이 다른 스레드에 의해 수정되었을 때 발생하는 예외
    - 동시 수정(concurrent modification)을 금지했는데, 발생해 버렸다는 것을 의미

        ```kotlin
        val list = mutableListOf(1, 2, 3)
        for (item in list) {
            if (item == 2) {
                list.remove(item)  // ConcurrentModificationException 발생 가능
            }
        }
        ```

      → iterator.remove()나 다른 안전한 방법을 사용해야 함

2. `UnsupportedOperationException`
    - 해당 객체가 지원하지 않는 메서드를 호출하려고 할 때 발생하는 예외
    - 사용자가 사용하려고 했던 메서드가 현재 객체에서는 사용할 수 없음을 나타냄

        ```kotlin
        val list = listOf(1, 2, 3)
        list.add(4)  // UnsupportedOperationException 발생
        ```

      → 기본적으로 사용할 수 없는 메서드는 클래스에 없는 것이 좋음 (인터페이스 분리 원칙 위반)


3. `NoSuchElementException`
    - 요청된 요소가 존재하지 않을 때 발생하는 예외
    - 사용자가 사용하려고 했던 요소가 존재하지 않음을 나타냄

        ```kotlin
        val list = emptyList<Int>()
        val firstElement = list.first()  // NoSuchElementException 발생
        ```

      → firstOrNull() 같은 메서드를 사용하여 요소가 없을 때 null을 반환하도록 할 수 있음



---



# [ 7. 결과 부족이 발생할 경우 null과 Failure를 사용하라 ]
https://marchbreeze.notion.site/7-null-Failure-51b93a0409da4e088618d4a2459dec21?pvs=4

- 함수가 원하는 결과를 만들어 낼 수 없을 때, 상황을 처리하는 매커니즘
    1. null 또는 '실패를 나타내는 sealed 클래스(일반적으로는 Failure)'를 리턴
    2. 예외를 throw

<aside>
💡 1. 충분히 예측할 수 있는 범위의 오류는 null, Failure로 처리
2. 예측하기 어려운 예외적인 범위의 오류는 예외를 throw해서 처리

</aside>

## 1. 예외

- `예외`는 정보를 전달하는 방법으로 사용해서는 안됨

    ```kotlin
    // 나쁜 예시
    fun calculateSum(a: Int, b: Int): Result<Int> = runCatching {
        if (a + b > 100) throw Exception("Sum is greater than 100")
        else a + b
    }
    ```

- 예외는 예외적인 상황이 발생했을 때 사용해야 함
- 이유 :
    1. 예외가 전파되는 과정을 제대로 추적하는 과정의 어려움
    2. 코틀린의 모든 예외는 unchecked 예외 → 사용자가 예외를 처리하지 않을 수도 있음
    3. 예외적인 상황을 처리하기 위해 만들어졌으므로, 명시적인 테스트(explicit test)만큼 빠르게 동작하지 않음
    4. try-catch 블록 안에 코드를 배치하면 컴파일러가 할 수 있는 최적화가 제한
        - try-catch 블록은 일반적인 실행 경로와 다른 제어 흐름을 가지고 있음
        - 예외가 발생할 경우에만 catch 블록이 실행 → 이러한 동적인 실행 경로 때문에 컴파일러가 최적화 하기 어려움

## 2. null & Failure

- `null`, `Failure`는 예상되는 오류를 표현할 때 효율적임
- 명시적이고 효율적이며 간단한 방법으로 처리 가능
- 추가적인 정보를 전달해야 한다면 sealed result 를 사용하고, 그렇지 않으면 null 을 사용하는 것이 일반적

- 공용체 타입(union type): 여러 다른 타입 중 하나를 가질 수 있는 타입 → Result와 같은 리턴방식 활용

    ```kotlin
    inline fun <reified T> String.readObjectOrNull(): T? {
        // ...
        if (incorrectSign) {
            return null
        }
        // ...
        return result
    }
    
    inline fun <reified T> String.readObject(): Result<T> {
        // ...
        if (incorrectSign) {
            return Failure(JsonParsingException())
        }
        // ...
        return Success(result)
    }
    
    sealed class Result<out T>
    class Success<out T>(val result: T): Result<T>()
    class Failure(val throwable: Throwable): Result<Nothing>()
    
    class JsonParsingException: Exception()
    ```


- null을 처리해야 한다면 사용자는 safe call 또는 엘비스 연산자 같은 다양한 null-safety 기능을 활용

    ```kotlin
    val age = userText.readObjectOrNull<Person>()?.age ?: -1
    ```

- 공용체(union type)를 리턴하기로 했다면 when 표현식을 써서 처리

    ```kotlin
    val person = userText.readObjectOrNull<Person>()
    val age = when(person) {
        is Success -> person.age
        is Failure -> -1
    }
    ```


- 이런 오류 처리 방식은 try-catch 블록보다 효율적이고 사용하기 쉽고 더 명확함
- 예외는 놓칠 수도 있고 전체 애플리케이션을 중지시킬 수도 있음

- 개발자는 항상 자신이 요소를 안전하게 추출할 거라 생각 → nullable을 리턴하면 안됨
- `getOrNull`을 써서 뭐가 리턴되는지 예측할 수 있게 하는 게 좋음
    - get : 특정 위치의 요소를 추출할 때 사용 → 해당 위치에 없으면 IndexOutOfBoundsException 발생
    - getOrNull : out of range 오류가 발생할 수 있는 경우 사용. 발생 시 null 리턴



# [ 8. 적절하게 null을 처리하라 ]
https://marchbreeze.notion.site/8-null-bf789936c4824a03b4a68e30487540d5?pvs=4



## 0. Null 처리방법

- null : 값이 부족하다(lack of value)
- 프로퍼티가 null = 값이 설정되지 않았거나 제거됐음을 나타냄
- 함수가 null 반환 = 함수에 따라 여러 의미를 가질 수 있음

    ```kotlin
    // String -> Int로 적절하게 변환할 수 없을 경우 null 리턴
    String.toIntOrNull()
    
    // 주어진 조건에 맞는 요소가 없을 경우 null 리턴
    Iterable<T>.firstOrNull(() -> Boolean)
    ```


- null은 최대한 명확한 의미를 가져야 함 (개발자가 nullable 값을 직접 처리해야하기 때문)

    ```kotlin
    val printer: Printer? = getPrinter()
    
    printer.print() // 컴파일 오류
    printer?.print() // 안전 호출
    printer!!.print()   // not-null assertion
    if (printer != null) printer.print() // 스마트 캐스팅
    ```


- nullable 타입의 처리방법 3가지 :
    1. safe call(?.), 스마트 캐스팅, 엘비스 연산자 등을 활용해서 안전하게 처리한다
    2. 오류를 throw한다
    3. 함수 or 프로퍼티를 리팩토링해서 nullable 타입이 안 나오게 바꾼다

## 1. null을 안전하게 처리하기

1. safe call & 스마트 캐스팅

    ```kotlin
    printer?.print() // safe call
    if (printer != null) printer.print() // 스마트 캐스팅
    ```

    - 앱 사용자 관점에서 가장 안전한 방법 (방어적 프로그래밍)

      ↔ 공격적 프로그래밍 (require, assert, check,…)


1. 엘비스 연산자

    ```kotlin
    val printerName1 = printer?.name ?: "Unnamed"
    val printerName2 = printer?.name ?: return
    val printerName3 = printer?.name ?: throw Error("Printer must be named")
    ```

    - 오른쪽에 return 또는 throw를 포함한 모든 표현식이 허용됨

1. 객체 자체의 확장 함수

    ```kotlin
    // 컬렉션 처리를 할 때 뭔가 없다는 걸 나타낼 때는 null이 아닌 빈 컬렉션을 쓰는 게 일반적
    Collection<T>.orEmpty()
    ```


## 2. 오류 throw하기

- null을 안전하게 처리하는 경우, 개발자에게 알리지 않고 코드가 진행됨 → 오류 찾기 어려워짐
- 오류를 강제 발생시킬 때는 throw, !!, requireNotNull, checkNotNull 등을 활용

    ```kotlin
    fun process(user: User) {
        requireNotNull(user.name)
        val context = checkNotNull(context)
        val networkService = getNetworkService(context) ?: throw NoInternetConnection()
        networkService.getData { data, userData -> 
            show(data!!, userData!!) 
        }
    }
    ```


- not-null assertion(!!) 문제
    - !!은 사용하기 쉽지만 좋은 해결법은 아님
    - 예외가 발생할 때 어떤 설명도 없는 제네릭 예외가 발생
    - 미래의 누군가가 함수를 리팩토링하면서 null일 수 있다는 걸 놓칠 경우, NPE가 발생함

      → 코드를 작성하는 올바른 방법은 lateinit 또는 Delegates.notNull을 쓰는 것


1. `lateinit 프로퍼티`
    - 프로퍼티가 이후에 설정될 것임을 명시하는 한정자
    - 초기화 전에 값을 쓰려고 하면 예외가 발생
    - 이후 어떤 의미를 나타내기 위해 null을 쓰고 싶을 때 nullable로 만들 수도 있음
    - 프로퍼티 초기화 이후에는 초기화되지 않은 상태로 돌아갈 수 없음

        ```kotlin
        class UserControllerTest {
            private lateinit var dao: UserDao?
            private lateinit var controller: UserController?
        
            @BeforeEach
            fun init() {
                dao = mockk()
                controller = UserController(dao)
            }
        
            @Test
            fun test() {
                controller.doSomething()
            }
        
        }
        ```


    - lateinit을 쓸 수 없는 경우: JVM에서 Int, Long, Double, Boolean 같은 기본형과 연결된 타입으로 프로퍼티를 초기화해야 하는 경우 → lateinit보다 느리지만 Deletages.notNull을 사용
        
        > lateinit은 기본형 타입(Int, Long, Double, Boolean 등)에는 사용할 수 없습니다. 이는 주로 lateinit이 nullable이 아닌 객체 타입에만 사용할 수 있도록 설계되었기 때문입니다.
        
        이유 1: 기본형 타입의 초기화 필요성
        기본형 타입은 항상 초기화되어야 하는 특성을 가지고 있습니다. 기본형 타입의 변수는 초기화되지 않은 상태로 사용할 수 없기 때문에 lateinit을 사용할 수 없습니다. 예를 들어, Int는 초기화되지 않은 상태로 남겨둘 수 없습니다.
        > 
        > 
        > **이유 2: 기본형 타입의 nullability**
        > 
        > lateinit은 null이 아닌 객체 참조 타입에만 사용되기 때문에 기본형 타입에는 사용할 수 없습니다. 기본형 타입은 null이 될 수 없고, 항상 값을 가져야 합니다. 반면에 lateinit은 null이 아닌 변수를 나중에 초기화할 것을 보장하기 위한 메커니즘입니다.
        > 
        > 이유 3: Delegates.notNull()을 사용하는 이유
        > 
        > 기본형 타입의 변수에 대해 lateinit과 유사한 지연 초기화를 사용하려면 Delegates.notNull()을 사용할 수 있습니다. 이 방법은 다음과 같은 경우에 유용합니다:
        > 
        > - 기본형 타입의 변수를 null로 초기화할 수 없는 상황에서 초기화 시점을 연기하고 싶을 때.
        > - lateinit과 유사하게 나중에 값을 할당할 수 있게 하고, 할당되지 않은 상태에서 사용하려고 하면 예외를 발생시켜 프로그램 오류를 방지하고자 할 때.
        > 
        > 예를 들어, Delegates.notNull()을 사용하면 초기값 없이 선언된 변수가 나중에 사용되기 전에 반드시 초기화되도록 강제할 수 있습니다. 그렇지 않으면 프로그램이 IllegalStateException을 발생시켜 초기화되지 않은 변수를 사용하는 오류를 방지할 수 있습니다.
        > 

2. `notNull Delegate`

    ```kotlin
    class DoctorActivity: Activity() {
        private var doctorId: Int by Delegates.notNull()
        private var fromNotification: Boolean by Delegates.notNull()
        
        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            doctorId = intent.extras.getInt(DOCTOR_ID_ARG)
            fromNotification = intent.extras.getBoolean(FROM_NOTIFICATION_ARG)
        }
    }
    ```


## 3. 의미 없는 nullability 피하기

- nullability는 어떻게든 적절하게 처리해야 해서 추가 비용이 발생 → 피하는게 좋음

- 클래스에서 nullability에 따라 여러 함수를 만들어 제공 → ex. List<T>, get, getOrNull()
- 값이 클래스 생성 이후에 확실하게 설정된다는 보장이 있으면 lateinit 프로퍼티와 notNull 델리게이트 사용
- 빈 컬렉션 대신 null을 리턴 X - 컬렉션을 빈 컬렉션으로 둘 때와 null로 둘 때는 의미가 다름
- nullable enum, None enum 값은 완전히 다른 의미 - null enum은 별도 처리해야 하지만 None enum 정의에 없으므로 필요한 경우에 사용하는 쪽에서 추가해서 활용할 수 있단 의미를 가짐



# [ 9. use를 사용하여 리소스를 닫아라 ]
https://marchbreeze.notion.site/9-use-af142e310115493b939520ad347eb30b?pvs=4




- 더 이상 필요하지 않을 때 close()를 써서 명시적으로 닫아야 하는 리소스가 존재함
    - Input/OutputStream
    - java.sql.Connection
    - java.io.Reader(FileReader, BufferedReader, CSSParser)
    - java.new.socket, java.util.Scanner

- 이런 리소스들은 AutoCloseable을 상속받는 Closeable 인터페이스를 구현함
    - 이런 모든 리소스는 최종적으로 리소스에 대한 레퍼런스가 없어질 때 가비지 컬렉터가 처리
    - 굉장히 느리고, 그 동안 리소스를 유지하는 비용이 많이 들어감

      ⇒  더 이상 필요없다면 명시적으로 close()를 호출하는 게 좋음


- 기존에서는 이런 리소스는 try-catch 블록을 써서 처리됨

    ```kotlin
    fun countCharactersInFile(path: String): Int {
        val reader = BufferedReader(FileReader(path))
        try {
            return reader.lineSequence().sumBy { it.length }
        }   finally {
            reader.close()
        }
    }
    ```

    - 복잡하고 좋지 않음 (예외 처리 로직이 없기 때문)
    - try 블록, finally 블록 안에서 오류가 발생하면 둘 중 하나만 전파됨 → use() 사용하는 것이 권장됨

- `use 함수`

    ```kotlin
    fun countCharactersInFile(path: String): Int {
        val reader = BufferedReader(FileReader(path))
        reader.use { 
            return reader.lineSequence().sumBy { it.length }
        }
    }
    ```

    ```kotlin
    // 람다 매개변수로 전달되는 형태 활용 시
    fun countCharactersInFile(path: String): Int {
        BufferedReader(FileReader(path)).use { reader ->
            return reader.lineSequence().sumBy { it.length }
        }
    }
    ```

    ```kotlin
    // 파일을 리소스로 쓰는 경우가 많고 파일을 한 줄씩 읽어들이는 경우
    fun countCharactersInFile(path: String): Int {
        File(path).useLines { lines ->
            return lines.sumBy { it.length }
        }
    }
    ```



# [ 10. 단위 테스트를 만들어라 ]
https://marchbreeze.notion.site/10-0510f075a65c400b9b185dd86e982823?pvs=4



- 코드가 올바르게 작동한다는 걸 보증하고 개발 시점에서의 빠른 피드백이 필요 : `단위 테스트`(unit test)

- 단위 테스트가 확인할 수 있는 상황
    - 일반적인 유스 케이스 (= happy path)
    - 일반적인 오류 케이스와 잠재적인 문제
    - 에지 케이스와 잘못된 아규먼트

- 단위 테스트의 장점
    - 테스트가 잘 된 코드는 신뢰할 수 있음
    - 테스트가 잘 만들어져 있다면, 리팩토링하는 것이 두렵지 않음 (버그가 생겼는지 쉽게 확인)
    - 수동으로 테스트하는 것보다 빠름

- 단위 테스트의 단점
    - 단위 테스트를 만드는 데 시간 소요 (물론 장기적으로는 디버깅 시간과 버그를 찾는 데 소모되는 시간을 줄여줌)
    - 테스트를 활용할 수 있게 코드를 수정해야 함 → 그래서 이런 변경을 통해 잘 정립된 아키텍처를 사용하도록 강제할 수 있음
    - 좋은 단위 테스트를 만드는 것은 어려움 (잘못 만들어진 단위 테스트는 득보다 실이 큼)

- 단위 테스트로 구현하면 좋은 부분
    - 복잡한 부분
    - 계속해서 수정이 일어나고 리팩터링이 일어날 수 있는 부분
    - 비즈니스 로직 부분
    - 공용 API 부분
    - 문제가 자주 발생하는 부분
    - 수정해야 하는 프로덕션 버그